Save and restore state in start TE activity

juanlaube opened the pull request:
Closes #2172 

I'm not convinced about this for multiple reasons:

- I wasn't able to properly test a tombstone scenario, even with a low ram emulator.
- I tested this by enabling `Don't keep activities` in the developer options and I had to add things in the Activity to make things work.

kornelijepetak commented on Toggl.Foundation.MvvmCross/ViewModels/StartTimeEntryViewModel.cs:
```diff
@@ -286,6 +294,43 @@ public override void Prepare(StartTimeEntryParameters parameter)
+            StartTime = startTime;
+
+            bool isBillable;
+            bool.TryParse(state.Data[StateBundleParams.IsBillable.ToString()], out isBillable);
```
`out var isBillable`

kornelijepetak commented on Toggl.Foundation.MvvmCross/ViewModels/StartTimeEntryViewModel.cs:
```diff
@@ -286,6 +294,43 @@ public override void Prepare(StartTimeEntryParameters parameter)
+            catch (JsonSerializationException) { }
+
+            DateTimeOffset startTime;
+            DateTimeOffset.TryParse(state.Data[StateBundleParams.StartTime.ToString()], out startTime);
```
`out var startTime`

kornelijepetak commented on Toggl.Foundation.MvvmCross/ViewModels/StartTimeEntryViewModel.cs:
```diff
@@ -286,6 +294,43 @@ public override void Prepare(StartTimeEntryParameters parameter)
+                var textFieldInfo = JsonConvert.SerializeObject(TextFieldInfo);
+                bundle.Data[StateBundleParams.TextFieldInfo.ToString()] = textFieldInfo;
+            } 
+            catch (JsonSerializationException) { }
```
If this fails, the exception is swallowed. That means the state you save to the bundle will not be complete and that reload from bundle won't be complete either, leading to inconsistent state of the app.

kornelijepetak commented on Toggl.Foundation.MvvmCross/ViewModels/StartTimeEntryViewModel.cs:
```diff
@@ -286,6 +294,43 @@ public override void Prepare(StartTimeEntryParameters parameter)
+
+            try {
+                long workspaceId;
+                long.TryParse(state.Data[StateBundleParams.WorkspaceId.ToString()], out workspaceId);
```
since C# 6, you can do `TryParse(..., out var workspaceId)`. No need to declare a variable before.

kornelijepetak commented on Toggl.Foundation.MvvmCross/ViewModels/StartTimeEntryViewModel.cs:
```diff
@@ -286,6 +294,43 @@ public override void Prepare(StartTimeEntryParameters parameter)
+                var textFieldInfo = JsonConvert.DeserializeObject<TextFieldInfo>(state.Data[StateBundleParams.TextFieldInfo.ToString()]);
+                TextFieldInfo = textFieldInfo.WithWorkspace(workspaceId);
+            }
+            catch (JsonSerializationException) { }
```
The same here. I am not sure whether we should allow this to just go unnoticed. 
When I was a kid, I was afraid of heights and darkness. Now, I am afraid of inconsistent states.

juanlaube commented:
And I'm also requesting @willsb review.

kornelijepetak commented on Toggl.Foundation.MvvmCross/Extensions/BundleExtensions.cs:
```diff
@@ -0,0 +1,118 @@
+                                textFieldInfo = textFieldInfo.WithProjectInfo(workspaceId, projectId, projectName, projectColor);
+                            }
+                        }
+                    } else
```
Else on new line

kornelijepetak commented on Toggl.Foundation/Autocomplete/TextFieldInfo.cs:
```diff
@@ -99,6 +99,16 @@ public TextFieldInfo AddTag(TagSuggestion tagSuggestion)
+        public TextFieldInfo AddTags(TagSuggestion[] tagSuggestions)
+        {
+            var result = this;
+            foreach (var tagSuggestion in tagSuggestions)
```
Since `.AddTag` creates a whole new `TextFieldInfo` object, I find this very unperformant. Similar to concatenating a lot of small strings with `+=`, there will be a lot array recreations and member copying...

I suggest creating another `TextFieldInfo` method: `AddTags` that would check for the existance of multiple tags and add them all in one batch and not recreate a `TFI` object in every step.

I like immutability, but this is really necessary IMO.

kornelijepetak commented on Toggl.Foundation.MvvmCross/Extensions/BundleExtensions.cs:
```diff
@@ -0,0 +1,118 @@
+                    {
+                        var tagIds = tagsIdsString.Split(';').Select(str => long.Parse(str)).ToArray();
+                        viewModel.TagIdsToReload = tagIds;
+                    } else {
```
do an early return instead.

kornelijepetak commented on Toggl.Foundation.MvvmCross/Extensions/BundleExtensions.cs:
```diff
@@ -0,0 +1,118 @@
+                            }
+                        }
+                    } else
+                    {
```
early return here to avoid that much branch nesting...

juanlaube commented:
@simonrozsival I've removed the JSON serialization code I added.
I was going that way first but @willsb didn't want to depend on json.net for this because it was supposed to be easy to serialize manually.

Narrator: it wasn't.

simonrozsival commented on Toggl.Foundation/Serialization/IJsonSerializer.cs:
```diff
@@ -2,7 +2,7 @@
 namespace Toggl.Foundation.Serialization
 {
-    internal interface IJsonSerializer
+    public interface IJsonSerializer
```
This could be reverted as well, right?

simonrozsival commented on Toggl.Foundation/Serialization/JsonSerializer.cs:
```diff
@@ -2,7 +2,7 @@
 namespace Toggl.Foundation.Serialization
 {
-    internal sealed class JsonSerializer : IJsonSerializer
+    public sealed class JsonSerializer : IJsonSerializer
```
This could be reverted as well, right?

juanlaube replied:
Right

kornelijepetak commented on Toggl.Foundation/Autocomplete/TextFieldInfo.cs:
```diff
@@ -101,12 +101,12 @@ public TextFieldInfo AddTag(TagSuggestion tagSuggestion)
-                result = result.AddTag(tagSuggestion);
-            }
-            return result;
+            var currentTagIds = Tags.Select(tag => tag.TagId);
```
I would do 

`var currentTagIds = new HashSet<long>(Tags.Select(tag => tag.TagId));`

here, to change the following `.Where` from `O(M x N)` into a `O(N)`

kornelijepetak commented on Toggl.Foundation.MvvmCross/Extensions/BundleExtensions.cs:
```diff
@@ -89,10 +89,15 @@ public static void ReloadPropertiesInto(this IMvxBundle bundle, StartTimeEntryVi
                             else
                             {
-                                textFieldInfo = textFieldInfo.WithProjectInfo(workspaceId, projectId, projectName, projectColor);
+                                return;
```
What I meant with early return is to revert the condition and return immediately so there wouldn't be an `else` at all.

kornelijepetak commented on Toggl.Foundation.MvvmCross/Extensions/BundleExtensions.cs:
```diff
@@ -89,10 +89,15 @@ public static void ReloadPropertiesInto(this IMvxBundle bundle, StartTimeEntryVi
+                            textFieldInfo = textFieldInfo.WithProjectInfo(workspaceId, projectId, projectName, projectColor);
+                        }
+                    }
+                    else
```
Same here.

kornelijepetak commented on Toggl.Foundation.MvvmCross/Extensions/BundleExtensions.cs:
```diff
@@ -104,7 +109,9 @@ public static void ReloadPropertiesInto(this IMvxBundle bundle, StartTimeEntryVi
                         viewModel.TagIdsToReload = tagIds;
-                    } else {
+                    }
+                    else
```
And here.

daividssilverio commented on Toggl.Foundation.MvvmCross/Extensions/BundleExtensions.cs:
```diff
@@ -0,0 +1,117 @@
+                        return;
+                    }
+                    var tagIds = tagsIdsString.Split(';').Select(str => long.Parse(str)).ToArray();
+                    viewModel.TagIdsToReload = tagIds;
```
I'm a little uneasy about this extension. Isn't modifying the VM from an extension of the MvxBundle a little bit of an inappropriate intimacy? It feels like the VM (or one extension for the VM) should be responsible for these methods.

Especially the ReloadPropertiesInto part, which is actually modifying stuff on the VM.

simonrozsival replied:
If this method returned some DTO which would then be used by the VM itself to update its state it would be nicer - agreed. The properties which were exposed could remain hidden.

juanlaube replied:
I like this idea, the extension doesn't need to know the internals of the view model

daividssilverio commented on Toggl.Foundation.MvvmCross/Extensions/BundleExtensions.cs:
```diff
@@ -0,0 +1,117 @@
+
+                if (hasTags)
+                {
+                    if (!(bundle.Data.TryGetValue($"{nameof(TextFieldInfo)}.{nameof(TextFieldInfo.Tags)}", out var tagsIdsString)))
```
there is one extra pair of unnecessary parenthesis here. after the `!`.

juanlaube commented:
@kornelijepetak I didn't have the time to review #2410 yet, but I plan to do it today :D

juanlaube commented:
Ping @willsb and @daividssilverio 😄

daividssilverio commented:
Other than comments, maybe we could something later about where those extensions are placed. Maybe a partial class would be nice? It feels like this something really specific for the `StartTimeEntryViewModel`.
This is a place where some "magic" would be cool. Haha.

(I'm not requesting changes for this now, just leaving my opinion).
